File: src/tui/siblings.ts
import type { Dirent } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";

async function isGitRepo(dir: string): Promise<boolean> {
  try {
    const stat = await fs.stat(path.join(dir, ".git"));
    return stat.isDirectory() || stat.isFile();
  } catch {
    return false;
  }
}

async function collectGitRepos(dir: string, exclude?: string): Promise<string[]> {
  let entries: Dirent[];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch {
    return [];
  }

  const results: string[] = [];
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const full = path.join(dir, entry.name);
    if (exclude && full === exclude) continue;
    if (await isGitRepo(full)) results.push(full);
  }
  return results;
}

export async function detectSiblings(cwd: string): Promise<string[]> {
  const resolved = path.resolve(cwd);

  if (await isGitRepo(resolved)) {
    const parent = path.dirname(resolved);
    if (parent === resolved) return [];
    return collectGitRepos(parent, resolved);
  }

  return collectGitRepos(resolved);
}

File: internal/tui/siblings.go
package tui

import (
	"os"
	"path/filepath"
)

// isGitRepo reports whether dir contains a .git entry (file or dir).
func isGitRepo(dir string) bool {
	info, err := os.Stat(filepath.Join(dir, ".git"))
	if err != nil {
		return false
	}
	return info.IsDir() || info.Mode().IsRegular()
}

// collectGitRepos returns immediate child directories of dir that are git
// repos, skipping exclude (when non-empty).
func collectGitRepos(dir, exclude string) []string {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	var out []string
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		full := filepath.Join(dir, e.Name())
		if exclude != "" && full == exclude {
			continue
		}
		if isGitRepo(full) {
			out = append(out, full)
		}
	}
	return out
}

// DetectSiblings returns absolute paths of git repos that are siblings of
// cwd (or, when cwd itself isn't a git repo, child repos of cwd). Mirrors
// src/tui/siblings.ts.
func DetectSiblings(cwd string) []string {
	resolved, err := filepath.Abs(cwd)
	if err != nil {
		return nil
	}
	if isGitRepo(resolved) {
		parent := filepath.Dir(resolved)
		if parent == resolved {
			return nil
		}
		return collectGitRepos(parent, resolved)
	}
	return collectGitRepos(resolved, "")
}
